File: leet/path_sum_3/path_sum_3.js
#!/usr/bin/env node
/*
437. Path Sum III
O(n) time, O(n) space solution.
*/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number}
 */

var count;
var seen;

var dfs = function(root, sum, target) {
    if (!root) return;
    sum += root.val;
    count += seen[sum - target] || 0;
    if (!seen[sum]) seen[sum] = 0;
    seen[sum]++;
    dfs(root.left, sum, target);
    dfs(root.right, sum, target);
    seen[sum]--;
}

var pathSum = function(root, sum) {
    count = 0;
    seen = {0: 1};
    dfs(root, 0, sum);
    return count;
};

File: leet/path_sum_3/path_sum_3.java
#!/usr/bin/env java
/*
437. Path Sum III
O(n) time, O(n) space solution.
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.HashMap;

class Solution {
    int count = 0;
    Map<Integer, Integer> seen = new HashMap<>();
    public void dfs(TreeNode root, int sum, int target)
    {
        if (root == null) return;
        sum += root.val;
        count += seen.getOrDefault(sum - target, 0);
        seen.put(sum, seen.getOrDefault(sum, 0) + 1);
        dfs(root.left, sum, target);
        dfs(root.right, sum, target);
        seen.put(sum, seen.get(sum) - 1);
    }
    public int pathSum(TreeNode root, int sum) {
        seen.put(0, 1);
        dfs(root, 0, sum);
        return count;
    }
}
